optimize_for_mobile: bring packed params to root module

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42041 optimize_for_mobile: bring packed params to root module**
* #42461 fix asan failure for fold_conv_bn when conv is quantized

Summary:

Adds a pass to hoist conv packed params to root module.
The benefit is that if there is nothing else in the conv module,
subsequent passes will delete it, which will reduce module size.

For context, freezing does not handle this because conv packed
params is a custom object.

Test Plan:

```
PYTORCH_JIT_LOG_LEVEL=">hoist_conv_packed_params.cpp" python test/test_mobile_optimizer.py TestOptimizer.test_hoist_conv_packed_params
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22801605](https://our.internmc.facebook.com/intern/diff/D22801605)

vkuzo committed:
optimize_for_mobile: bring packed params to root module

Summary:

Adds a pass to hoist conv packed params to root module.
The benefit is that if there is nothing else in the conv module,
subsequent passes will delete it, which will reduce module size.

For context, freezing does not handle this because conv packed
params is a custom object.

Test Plan:

```
PYTORCH_JIT_LOG_LEVEL=">hoist_conv_packed_params.cpp" python test/test_mobile_optimizer.py TestOptimizer.test_hoist_conv_packed_params
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo commented on torch/csrc/jit/passes/hoist_conv_packed_params.cpp:
```diff
@@ -0,0 +1,129 @@
+
+    for (Node* n : b->nodes()) {
+
+      // TODO before land: also check for n->inputs()[0]->node()->kind() == {ConvNd}
```
would love some tips on best way to do this

jerryzh168 replied:
please checkout: https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/passes/quantization/helper.cpp.html#_ZN5torch3jit16is_conv2d_moduleERKNS0_5MatchERKSt13unordered_mapINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEPNS0_5ValueESt4hashISA_ESt8equal_11899141

kimishpatel commented on torch/csrc/jit/passes/xnnpack_rewrite.cpp:
```diff
@@ -294,6 +296,11 @@ script::Module optimizeForMobile(
 
+  if (!optimization_blocklist.count(
+      MobileOptimizerType::HOIST_CONV_PACKED_PARAMS)) {
+    HoistConvPackedParams(cloned_module);
```
I am assuming this requires inlining? If so then you might want to run freeze_module once again? Since other pass might be disabled?

jerryzh168 replied:
yeah we may need to freeze again to remove the conv module itself I think

kimishpatel replied:
I also meant before. Doesnt this pass require inlining of the graph as well? Because right now inlining is done by freeze_module, and if you have INSERT_FOLD_PREPACK_OPS blocklisted, it wont do that. So if inlining is required I suggest you call Inline explicitly before calling `HoistConvPackedParams`

jerryzh168 replied:
yeah, I have the same question, we need to make sure what is required before running this pass. I think inline is required in this case. Also actually I'm not sure if it is OK to freeze module for user if it is not frozen previously.

jerryzh168 replied:
@kimishpatel  In general I feel this is an optimization only valid when freeze is enabled, is there a way to enforce that in `optimization_blocklist`?

vkuzo replied:
sounds good.  We can throw an assert here for `INSERT_FOLD_PREPACK_OPS` being enabled, unless there is a better way.

kimishpatel replied:
> @kimishpatel In general I feel this is an optimization only valid when freeze is enabled. why is freezing not an optimization option actually?

@jerryzh168, I think freezing was never thought of as an optimization think when I did this. Your point is fair, maybe we can add that.

@vkuzo, I would say instead of throwing an assert, we just add freezing before this. I understand it may mean we are calling freezing multiple times, but that should not be an issue. Plus that makes this pass self contained without any dependency on having to run another pass.

kimishpatel reviewed (changes requested):
Looks pretty good. Left one comment.

jerryzh168 commented on test/test_mobile_optimizer.py:
```diff
@@ -207,5 +208,88 @@ def get_lint_count_by_type(lint_type, module_lint_List):
+
+        # generic case
+
+        m = Parent()
```
nit: can you put these two test cases in a loop?

jerryzh168 replied:
and conv module attribute is removed

jerryzh168 commented on torch/csrc/jit/passes/hoist_conv_packed_params.cpp:
```diff
@@ -0,0 +1,129 @@
+  // create a vector of the getAttr nodes
+  //   idx 0 is the node from root to child 1
+  //   idx n-1 is the node from child n-1 to conv
+  std::vector<Node*> getAttrNodes;
+  Node* curNode = getConvPackedParamsNode->inputs()[0]->node();
+  while (!(curNode->outputs()[0]->type() == graph->inputs()[0]->type())) {
+    TORCH_CHECK(
+      curNode->kind() == prim::GetAttr,
+      "Attempted to add a non-prim::GetAttr node to a chain of prim::getAttr nodes.");
+    getAttrNodes.insert(getAttrNodes.begin(), curNode);
+    curNode = curNode->inputs()[0]->node();
+  }
+  TORCH_CHECK(getAttrNodes.size() > 0, "Did not find a chain of prim::getAttr nodes");
```
we have some helper function in quantization to do this: https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/passes/quantization/helper.cpp.html#_ZN5torch3jit16getInvokedModuleERNS0_6ModuleEPNS0_4NodeEPNS0_5ValueE, feel free to use them

jerryzh168 replied:
oh this one: https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/passes/quantization/helper.cpp.html#_ZN5torch3jit19getInvokedModuleOptERKNS0_6ModuleEPNS0_4NodeEPNS0_5ValueE since maybe we will fail to get module

jerryzh168 commented on torch/csrc/jit/passes/hoist_conv_packed_params.cpp:
```diff
@@ -0,0 +1,129 @@
+  for (int idx = 1; idx < getAttrNodes.size(); idx++) {
+    curName = getAttrNodes[idx]->s(attr::name);
+    curConvModule = curConvModule.attr(curName).toModule();
+  }
```
all code above can be achieved by getInvokedModule and getModuleAccessPath I think

jerryzh168 commented on torch/csrc/jit/passes/hoist_conv_packed_params.cpp:
```diff
@@ -0,0 +1,129 @@
+  std::string newName = prefix + "." + suffix + "_packed_params";
+
+  // make sure the attribute does not already exist
+  TORCH_CHECK(
```
if the name already exists maybe we can create a new one? e.g. https://codebrowser.bddppq.com/pytorch/pytorch/torch/csrc/jit/passes/quantization/insert_observers.cpp.html#937

vkuzo replied:
I would imagine if we hit a duplicate it would be more likely that this code has a bug, compared to the user having something named `_jit_pass_hoist_conv_packed_params.{foo}.{bar}.{conv_foobar}` before this pass.

jerryzh168 replied:
if `.` can appear in attribute names then we might hit duplicate in extreme cases, although it is very unlikely
e.g. consider {foo}.{bar} is also an attribute itself

vkuzo replied:
good point, thx, will update

jerryzh168 commented on torch/csrc/jit/passes/hoist_conv_packed_params.cpp:
```diff
@@ -0,0 +1,129 @@
+  // traverse the chain to get packed params value
+  std::string curName = getAttrNodes[0]->s(attr::name);
+  Module curConvModule = rootModule.attr(curName).toModule();
+  for (int idx = 1; idx < getAttrNodes.size(); idx++) {
```
lint

jerryzh168 commented on test/test_mobile_optimizer.py:
```diff
@@ -207,5 +208,88 @@ def get_lint_count_by_type(lint_type, module_lint_List):
+        m = torch.jit.script(m)
+        m_optim = optimize_for_mobile(m)
+
+        FileCheck().check_not("Conv2d = prim::GetAttr[name=\"conv1\"]") \
```
also can we check the access to conv module is removed?

vkuzo replied:
sure, will add

vkuzo replied:
actually, sorry, what do you mean specifically?  The original access is 
```
__torch__.torch.nn.quantized.modules.conv.___torch_mangle_13.Conv2d = prim::GetAttr[name="conv1"](%self)
```

I'm checking for the absence of a suffix of that (for readability).  Happy to adjust if something else is better

jerryzh168 replied:
I mean check the conv attribute is removed from the parent module

vkuzo replied:
makes sense.  Any tips on the best way to do that with `FileCheck`?

jerryzh168 replied:
FileCheck is just checking the strings. for this one we need to check the module itself with something like
`self.assertFalse(hasattr(module, 'conv1'))`

vkuzo committed:
Update on "optimize_for_mobile: bring packed params to root module"


Summary:

Adds a pass to hoist conv packed params to root module.
The benefit is that if there is nothing else in the conv module,
subsequent passes will delete it, which will reduce module size.

For context, freezing does not handle this because conv packed
params is a custom object.

Test Plan:

```
PYTORCH_JIT_LOG_LEVEL=">hoist_conv_packed_params.cpp" python test/test_mobile_optimizer.py TestOptimizer.test_hoist_conv_packed_params
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "optimize_for_mobile: bring packed params to root module"


Summary:

Adds a pass to hoist conv packed params to root module.
The benefit is that if there is nothing else in the conv module,
subsequent passes will delete it, which will reduce module size.

For context, freezing does not handle this because conv packed
params is a custom object.

Test Plan:

```
PYTORCH_JIT_LOG_LEVEL=">hoist_conv_packed_params.cpp" python test/test_mobile_optimizer.py TestOptimizer.test_hoist_conv_packed_params
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
comments on "optimize_for_mobile: bring packed params to root module"


Summary:

Adds a pass to hoist conv packed params to root module.
The benefit is that if there is nothing else in the conv module,
subsequent passes will delete it, which will reduce module size.

For context, freezing does not handle this because conv packed
params is a custom object.

Test Plan:

```
PYTORCH_JIT_LOG_LEVEL=">hoist_conv_packed_params.cpp" python test/test_mobile_optimizer.py TestOptimizer.test_hoist_conv_packed_params
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
clang on "optimize_for_mobile: bring packed params to root module"


Summary:

Adds a pass to hoist conv packed params to root module.
The benefit is that if there is nothing else in the conv module,
subsequent passes will delete it, which will reduce module size.

For context, freezing does not handle this because conv packed
params is a custom object.

Test Plan:

```
PYTORCH_JIT_LOG_LEVEL=">hoist_conv_packed_params.cpp" python test/test_mobile_optimizer.py TestOptimizer.test_hoist_conv_packed_params
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22801605](https://our.internmc.facebook.com/intern/diff/D22801605)

[ghstack-poisoned]

vkuzo commented on torch/csrc/jit/passes/xnnpack_rewrite.cpp:
```diff
@@ -294,6 +296,15 @@ script::Module optimizeForMobile(
+  if (!optimization_blocklist.count(
+          MobileOptimizerType::HOIST_CONV_PACKED_PARAMS)) {
+    // freeze again in case it was not done in previous optional passes
+    cloned_module = freeze_module(cloned_module, preserved_methods);
```
I'm seeing optimize_for_mobile test fail on this line (https://gist.github.com/vkuzo/0e0562fd46a10b9ed8f2f6e8eefca349), just checking that this is expected and we can update the test to match?  seems like that test case currently assumes the BN::forward call survives

kimishpatel replied:
No I dont think so.  BNTestModule has BN and Conv2d. The test where it fails, it has disabled prepack folding. If that is disabled there is no inlining, via freeze_module. So we should still have Conv2d::forward. I think that is what it is expecting. But since you have freeze_module after, and you probably will add freeze_module before, inlining will happen and there wont be any forward calls. So yeah we will have to fix that test somehow. Maybe best to remove that portion of the test.

jerryzh168 reviewed (approved):
LGTM, please fix the failing tests.

vkuzo committed:
fix test on "optimize_for_mobile: bring packed params to root module"


Summary:

Adds a pass to hoist conv packed params to root module.
The benefit is that if there is nothing else in the conv module,
subsequent passes will delete it, which will reduce module size.

For context, freezing does not handle this because conv packed
params is a custom object.

Test Plan:

```
PYTORCH_JIT_LOG_LEVEL=">hoist_conv_packed_params.cpp" python test/test_mobile_optimizer.py TestOptimizer.test_hoist_conv_packed_params
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22801605](https://our.internmc.facebook.com/intern/diff/D22801605)

[ghstack-poisoned]

vkuzo commented:
we should wait for https://github.com/pytorch/pytorch/issues/42462 to be resolved before landing this

vkuzo committed:
rebase on "optimize_for_mobile: bring packed params to root module"


Summary:

Adds a pass to hoist conv packed params to root module.
The benefit is that if there is nothing else in the conv module,
subsequent passes will delete it, which will reduce module size.

For context, freezing does not handle this because conv packed
params is a custom object.

Test Plan:

```
PYTORCH_JIT_LOG_LEVEL=">hoist_conv_packed_params.cpp" python test/test_mobile_optimizer.py TestOptimizer.test_hoist_conv_packed_params
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22801605](https://our.internmc.facebook.com/intern/diff/D22801605)

[ghstack-poisoned]

vkuzo committed:
lint on "optimize_for_mobile: bring packed params to root module"


Summary:

Adds a pass to hoist conv packed params to root module.
The benefit is that if there is nothing else in the conv module,
subsequent passes will delete it, which will reduce module size.

For context, freezing does not handle this because conv packed
params is a custom object.

Test Plan:

```
PYTORCH_JIT_LOG_LEVEL=">hoist_conv_packed_params.cpp" python test/test_mobile_optimizer.py TestOptimizer.test_hoist_conv_packed_params
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22801605](https://our.internmc.facebook.com/intern/diff/D22801605)

[ghstack-poisoned]

raghuramank100 commented on torch/csrc/jit/passes/hoist_conv_packed_params.cpp:
```diff
@@ -0,0 +1,128 @@
+// %n =
+// prim::GetAttr[name="{prefix}.name1{...}.name(n-1)._packed_params"][%self]
+//
+void hoistConvPackedParams(
```
Can we also add a pass for linear? Is there value in doing this for linear modules too?

vkuzo replied:
I don't see why not, although I'd defer to the mobile team on the priority.

raghuramank100 commented on test/test_mobile_optimizer.py:
```diff
@@ -248,8 +246,86 @@ def forward(self, x):
         model_optim = optimize_for_mobile(model)
-        self.assertFalse(hasattr(model_optim.conv1, "bias"))
-        self.assertFalse(hasattr(model_optim.child.conv2, "bias"))
+
```
Is the understanding that one cannot serialize/deserialize a model after optimize_for_mobile is called?

vkuzo committed:
add qnnpack check on "optimize_for_mobile: bring packed params to root module"


Summary:

Adds a pass to hoist conv packed params to root module.
The benefit is that if there is nothing else in the conv module,
subsequent passes will delete it, which will reduce module size.

For context, freezing does not handle this because conv packed
params is a custom object.

Test Plan:

```
PYTORCH_JIT_LOG_LEVEL=">hoist_conv_packed_params.cpp" python test/test_mobile_optimizer.py TestOptimizer.test_hoist_conv_packed_params
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22801605](https://our.internmc.facebook.com/intern/diff/D22801605)

[ghstack-poisoned]

vkuzo committed:
fix tests and add convrelu on "optimize_for_mobile: bring packed params to root module"


Summary:

Adds a pass to hoist conv packed params to root module.
The benefit is that if there is nothing else in the conv module,
subsequent passes will delete it, which will reduce module size.

For context, freezing does not handle this because conv packed
params is a custom object.

Test Plan:

```
PYTORCH_JIT_LOG_LEVEL=">hoist_conv_packed_params.cpp" python test/test_mobile_optimizer.py TestOptimizer.test_hoist_conv_packed_params
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22801605](https://our.internmc.facebook.com/intern/diff/D22801605)

[ghstack-poisoned]

kimishpatel reviewed (approved):
Looks good. Thanks.

vkuzo committed:
rebase on "optimize_for_mobile: bring packed params to root module"


Summary:

Adds a pass to hoist conv packed params to root module.
The benefit is that if there is nothing else in the conv module,
subsequent passes will delete it, which will reduce module size.

For context, freezing does not handle this because conv packed
params is a custom object.

Test Plan:

```
PYTORCH_JIT_LOG_LEVEL=">hoist_conv_packed_params.cpp" python test/test_mobile_optimizer.py TestOptimizer.test_hoist_conv_packed_params
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D22801605](https://our.internmc.facebook.com/intern/diff/D22801605)

[ghstack-poisoned]

vkuzo commented:
resubmitted as https://github.com/pytorch/pytorch/pull/42740 due to a ghstack snafu
